### PR TITLE
Remove predicate in IcebergTableHandle to clarify the responsibilities

### PR DESCRIPTION
**Pull request by hantangwangd**
## Description

Maintaining the predicate in multiple places by different code paths currently is quite confusing.

This PR totally discard the predicate in `IcebergTableHandle`, and just maintain the information about predicate in `IcebergTableLayoutHandle` inside iceberg connector. So that the responsibilities of `IcebergTableHandle`, `IcebergTableLayoutHandle` and `ConnectorTableLayout` could be clarified more clearer, and the behaviors could be more consistent between iceberg.pushdown_filter_enabled on and off.

## Motivation and Context

Make Iceberg connector implementation more clearer

## Impact

N/A

## Test Plan

 - Make sure the refactor do not affect existing test cases
 - Make sure the simplification of predicate on partitioned columns do not lead to incorrect result

## Contributor checklist

- [x] Please make sure your submission complies with our [development](https://github.com/prestodb/presto/wiki/Presto-Development-Guidelines#development), [formatting](https://github.com/prestodb/presto/wiki/Presto-Development-Guidelines#formatting), [commit message](https://github.com/prestodb/presto/wiki/Review-and-Commit-guidelines#commit-formatting-and-pull-requests), and [attribution guidelines](https://github.com/prestodb/presto/wiki/Review-and-Commit-guidelines#attribution).
- [x] PR description addresses the issue accurately and concisely.  If the change is non-trivial, a GitHub Issue is referenced.
- [x] Documented new properties (with its default value), SQL syntax, functions, or other functionality.
- [x] If release notes are required, they follow the [release notes guidelines](https://github.com/prestodb/presto/wiki/Release-Notes-Guidelines).
- [x] Adequate tests were added if applicable.
- [x] CI passed.

## Release Notes

```
== NO RELEASE NOTE ==
```

### CODE REVIEW

File: presto-iceberg/src/main/java/com/facebook/presto/iceberg/IcebergAbstractMetadata.java [hunk 1]
```diff
@@ -224,12 +224,12 @@ public ConnectorTableLayoutResult getTableLayoutForConstraint(
                 new IcebergTableLayoutHandle.Builder()
                         .setPartitionColumns(ImmutableList.copyOf(getPartitionKeyColumnHandles(icebergTable, typeManager)))
                         .setDataColumns(toHiveColumns(icebergTable.schema().columns()))
-                        .setDomainPredicate(constraint.getSummary().transform(IcebergAbstractMetadata::toSubfield))
+                        .setDomainPredicate(constraint.getSummary().simplify().transform(IcebergAbstractMetadata::toSubfield))
```
Comment (hantangwangd) [inline presto-iceberg/src/main/java/com/facebook/presto/iceberg/IcebergAbstractMetadata.java:227]: We should be careful about the `simplify()` here, for Iceberg could not support a large value list for IN predicate. So the `unenforcedConstraint` in the returned `ConnectorTableLayoutResult` should be `constraint.getSummary()`.

**Comment by tdcmeehan** [comment]
Are the protocol changes intentional?

**Comment by hantangwangd** [comment]
> Are the protocol changes intentional?

@tdcmeehan yes, `IcebergTableHandle` is defined in `presto_protocol.yml`, so we have to re-generate the c++ protocol code after we dropped it's `predicate` field.

Besides, the addition of `ConnectorHistogram` in `presto_protocol.yml` is to ensure that the generated c++ code could be correctly compiled. For in PR #22042 we added a field of `ConnectorHistogram` type in `VariableStatsEstimate` which was defined in `presto_protocol.yml`, so we have to define `ConnectorHistogram ` as well. @ZacBlanco is that OK?

The addition of `ConnectorHistogram` is duplicated with #22080 , currently they have to do this to pass the CI tests, and the latter merged one would modify it's `presto_protocol.yml`.

File: presto-iceberg/src/main/java/com/facebook/presto/iceberg/TableStatisticsMaker.java [hunk 2]
```diff
@@ -121,8 +121,7 @@ private TableStatistics makeTableStatistics(IcebergTableHandle tableHandle, Cons
         }
 
         TupleDomain<IcebergColumnHandle> intersection = constraint.getSummary()
-                .transform(IcebergColumnHandle.class::cast)
-                .intersect(tableHandle.getPredicate());
```
Comment (imjalpreet) [inline presto-iceberg/src/main/java/com/facebook/presto/iceberg/TableStatisticsMaker.java:125]: Is removing this intersection fine or should we pass the valid predicate from layout?

**Review by imjalpreet — COMMENTED** [review]
Mostly LGTM, just a small question in TableStatisticsMaker. 

Also, can we add a separate commit for the native protocol changes in a new PR. I believe that's the recommended way to merge the protocol changes as a separate PR.

File: presto-native-execution/presto_cpp/presto_protocol/presto_protocol.cpp
Comment (imjalpreet) [inline presto-native-execution/presto_cpp/presto_protocol/presto_protocol.cpp:1]: Also, just wanted to confirm that there have been no manual changes in the protocol files and everything is generated by the scripts?

File: presto-iceberg/src/main/java/com/facebook/presto/iceberg/TableStatisticsMaker.java (hunk 2 above)
Comment (hantangwangd) [reply presto-iceberg/src/main/java/com/facebook/presto/iceberg/TableStatisticsMaker.java:125]: Thanks for the reminder. Yes, I think we should pass the valid predicate from layout or merge it into the `constraint` from outer this method. Will fix it.

File: presto-native-execution/presto_cpp/presto_protocol/presto_protocol.cpp
Comment (hantangwangd) [reply presto-native-execution/presto_cpp/presto_protocol/presto_protocol.cpp:1]: Yes, it is completely generated.

File: presto-iceberg/src/main/java/com/facebook/presto/iceberg/IcebergTableHandle.java [hunk 3]
```diff
@@ -29,6 +29,7 @@ public class IcebergTableHandle
         extends BaseHiveTableHandle
 {
     private final IcebergTableName icebergTableName;
+    // TODO: this field is no longer useful, would be removed in a subsequent PR
```
Comment (ZacBlanco) [inline presto-iceberg/src/main/java/com/facebook/presto/iceberg/IcebergTableHandle.java:32]: Shouldn't we just remove in this PR? Or will it create a very large change? I feel like it should just be removed.

File: presto-iceberg/src/main/java/com/facebook/presto/iceberg/optimizer/IcebergEqualityDeleteAsJoin.java [hunk 4]
```diff
@@ -316,7 +322,7 @@ private TableScanNode createDeletesTableScan(ImmutableMap<VariableReferenceExpre
                             tableName.getSnapshotId(),
                             Optional.empty()),
                     icebergTableHandle.isSnapshotSpecified(),
-                    icebergTableHandle.getPredicate(),
+                    TupleDomain.all(),
```
Comment (ZacBlanco) [inline presto-iceberg/src/main/java/com/facebook/presto/iceberg/optimizer/IcebergEqualityDeleteAsJoin.java:325]: What's the reason for this change? Will the query result be wrong if the table is filtered on the predicate?

File: presto-iceberg/src/main/java/com/facebook/presto/iceberg/optimizer/IcebergEqualityDeleteAsJoin.java [hunk 5]
```diff
@@ -344,7 +350,7 @@ private TableScanNode createNewRoot(TableScanNode node, IcebergTableHandle icebe
                             tableName.getSnapshotId(),
                             tableName.getChangelogEndSnapshot()),
                     icebergTableHandle.isSnapshotSpecified(),
-                    icebergTableHandle.getPredicate(),
+                    TupleDomain.all(),
```
Comment (ZacBlanco) [inline presto-iceberg/src/main/java/com/facebook/presto/iceberg/optimizer/IcebergEqualityDeleteAsJoin.java:353]: same comment

**Review by ZacBlanco — COMMENTED** [review]
Just a few things. Looks good mostly

File: presto-iceberg/src/main/java/com/facebook/presto/iceberg/optimizer/IcebergEqualityDeleteAsJoin.java (hunk 4 above)
Comment (hantangwangd) [reply presto-iceberg/src/main/java/com/facebook/presto/iceberg/optimizer/IcebergEqualityDeleteAsJoin.java:325]: As the predicate is useless and would be dropped subsequently, but the native worker need this value to be not null, so for safety I just set it to `ALL`.

File: presto-iceberg/src/main/java/com/facebook/presto/iceberg/optimizer/IcebergEqualityDeleteAsJoin.java (hunk 5 above)
Comment (hantangwangd) [reply presto-iceberg/src/main/java/com/facebook/presto/iceberg/optimizer/IcebergEqualityDeleteAsJoin.java:353]: The same as above, just to show that it is useless, but could not be null for now.

**Review by ZacBlanco — DISMISSED** [review]

**Review by imjalpreet — DISMISSED** [review]
LGTM, thank you for changes!

**Review by tdcmeehan — APPROVED** [review]
